### PR TITLE
add first interface dummy

### PR DESCRIPTION
**Pull request by f-demirci**
This PR provides a first API for the robot test facility.

### CODE REVIEW

**Comment by f-demirci** [comment]
Realization of issue #4

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    """ API for controling the PILZ robot test facility """
+      
+    
+    def chooseOperationMode(self, op_mode):
+        """ Allows to change the operation mode """
+        raise NotImplementedError
+
+    def activateEmergency(self):
+        """ Triggers emergency stop by releases the STO_A und STO_B (STO-Safe Torque Off)"""
+        raise NotImplsementedError
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]:
```suggestion
        raise NotImplementedError
```

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+class TestFacilityControlAPI(Object):
+    """ API for controling the PILZ robot test facility """
+      
+    
+    def chooseOperationMode(self, op_mode):
+        """ Allows to change the operation mode """
+        raise NotImplementedError
+
+    def activateEmergency(self):
+        """ Triggers emergency stop by releases the STO_A und STO_B (STO-Safe Torque Off)"""
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]:
```suggestion
        """ Triggers emergency stop by releasing the STO_A und STO_B (STO-Safe Torque Off)"""
```

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        raise NotImplsementedError
+
+    def disableEmergency(self):
+        """ Realeses emergency stop by triggering the STO_A und STO_B (STO-Safe Torque Off) """
+        raise NotImplementedError
+
+    def acknowledgeReadySignal(self):
+        """ Acknowledges the rady signal of the safety controller """
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]:
```suggestion
        """ Acknowledges the ready signal of the safety controller """
```

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        raise NotImplementedError
+
+    def activateEnabling(self):
+        """ Enables the drives by triggering STO_A und STO_B (STO-Safe Torque Off) """
+        raise NotImplementedError
+
+    def disableEnabling(self):
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]: I do not know, what has been discussed, but I do not like the name `disableEnabling`.
Possible alternatives: deactivateEnabling? or simply enable/disable?

Comment (martiniil) [reply pilz_robot_tfc_api/src/tfc_api.py]: The docus for `activateEnabling` and `disableEmergency` are identical. So one of the functions is redundant. Or the docu is not sufficient.

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+"""API for easy usage of Pilz robot test facility."""
+
+OP_MODE_T1 = 0
+OP_MODE_T2 = 1
+OP_MODE_AUTO = 2
+
+
+class TestFacilityControlAPI(Object):
+    """ API for controling the PILZ robot test facility """
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]:
```suggestion
    """ API for controlling the PILZ robot test facility """
```

File: pilz_robot_tfc_api/src/tfc_api.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""API for easy usage of Pilz robot test facility."""
+
+OP_MODE_T1 = 0
+OP_MODE_T2 = 1
+OP_MODE_AUTO = 2
```
Comment (martiniil) [inline pilz_robot_tfc_api/src/tfc_api.py]: Please add documentation or name the constants in a way that makes clear, what they should be used for.

**Review by martiniil — CHANGES_REQUESTED** [review]
I think the API docu should be included in the ReadMe. Do you have plans to do so?

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 1 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Changed => Not needed anymore.

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 2 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Done.

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 3 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Changed => Not needed anymore.

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 4 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: > I do not know, what has been discussed, but I do not like the name `disableEnabling`.
> Possible alternatives: deactivateEnabling? or simply enable/disable?

Changed to "deactive_enabling"

Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: > The docus for `activateEnabling` and `disableEmergency` are identical. So one of the functions is redundant. Or the docu is not sufficient.

Yeah, that's a tricky one. The problem is: We want to "emulate" the real hardware setup. The real hardware setup consists of a enabling button and an emergency stop. Both do (if I'm correct) basically the same, hence, the same docu. However, for a correct test we need both functionalities. Therefore, I opt to leave both functions.

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 6 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Changed to enum => Done.

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 5 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Done.

**Comment by HiddenBug** [comment]
@martiniil Sorry, you were much faster with your review then I anticipated. I refactored the API base class to now use the ABCMeta class which allows to enforce that derived classes need to implement the abstract functions and that the base class cannot be instantiated. I also tried to clean up the package and CMakeList files and added missing files.

**Comment by HiddenBug** [comment]
> I think the API docu should be included in the ReadMe. Do you have plans to do so?

We have not talked about this yet. What do you suggest? Simply stating that there exists a API for controlling the robot test facility and referencing the sphinx docu?

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 4 above)
Comment (martiniil) [reply pilz_robot_tfc_api/src/tfc_api.py]: Yes, I understand, that it makes sense to keep both. But:
```
by triggering STO_A und STO_B
```
I think this is too specific for the general API. From this point of view it is not necessary that both functions do the same. This just follows from the actual implementation.

**Comment by martiniil** [comment]
> @martiniil Sorry, you were much faster with your review then I anticipated. I refactored the API base class to now use the ABCMeta class which allows to enforce that derived classes need to implement the abstract functions and that the base class cannot be instantiated. I also tried to clean up the package and CMakeList files and added missing files.

Sounds good.

**Comment by martiniil** [comment]
> > I think the API docu should be included in the ReadMe. Do you have plans to do so?
> 
> We have not talked about this yet. What do you suggest? Simply stating that there exists a API for controlling the robot test facility and referencing the sphinx docu?

I thought about directly including the API in the ReadMe, that is basically show the contents of tfc_api.py. What do you think?

File: pilz_robot_tfc_api/src/tfc_api.py (hunk 4 above)
Comment (HiddenBug) [reply pilz_robot_tfc_api/src/tfc_api.py]: Done. -> Improved API documentation.

**Comment by HiddenBug** [comment]
For simplicity I close this PR. PR #13 includes this PR and, therefore, continues the development started here.

**Comment by hslusarek** [comment]
> > > I think the API docu should be included in the ReadMe. Do you have plans to do so?
> > 
> > 
> > We have not talked about this yet. What do you suggest? Simply stating that there exists a API for controlling the robot test facility and referencing the sphinx docu?
> 
> I thought about directly including the API in the ReadMe, that is basically show the contents of tfc_api.py. What do you think?

I remember trying it like you suggested, in other words, to include the API in the README. But currently, this does not seem to work. It also does not work in the [PilzDe/pilz_industrial_motion/pilz_robot_pogramming repository](https://github.com/PilzDE/pilz_industrial_motion/tree/melodic-devel/pilz_robot_programming#the-code-demo_programpy)
Therefore, I suggest to skip this point for now and to do it as soon as a solution for the described problem is found.

**Pull request closed without merging**